### PR TITLE
Using Distributed Map state to increase concurrency processing for the web crawler.

### LINKED ISSUE #24: Using Distributed Map state to speed up the crawling process for large websites
@cogwirrel, first of all thank you for the great blog post and the sample code base for this Serverless web crawler stack. I have used this solution as an alternative to the Kendra web crawling connector which failed to crawl certain websites with unclear reasons.

When crawling large websites of tens or hundreds of thousands of links, the process took a long time due to the Inline Map state limit of 40 concurrent iterations. As discussed, I'm proposing to change the code base to use the new Distributed Map state that can support a maximum concurrency of up to 10,000 executions in parallel to speed up the crawling process.

### PR DESCRIPTION
**Pull request by philiptran**
When crawling large websites of tens or hundreds of thousands of links, the process took a long time due to the Inline Map state limit of 40 concurrent iterations. This PR uses the new Distributed Map state that can support a maximum concurrency of up to 10,000 executions in parallel to speed up the crawling process.

Fixes #24:

* Updated the state machine to use Distributed Map state and created S3 working bucket and required permissions.
* Allowed STATE_MACHINE_URL_THRESHOLD and PARALLEL_URLS_TO_SYNC to be configured via env variables.
* Saved the queuePaths to S3 bucket instead of passing the big array as part of the payload to avoid hitting the max request size limit of 256KB.
* Added explanation for setting the STATE_MACHINE_URL_THRESHOLD and PARALLEL_URLS_TO_SYNC to increase concurrency processing for the web crawler.

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

File: src/constructs/webcrawler/web-crawler-state-machine.ts [hunk 1]
```diff
@@ -1,15 +1,20 @@
...
 // SPDX-License-Identifier: MIT-0
 import { Construct } from 'constructs';
-import { Choice, Succeed, Map, StateMachine, Condition, TaskInput } from "aws-cdk-lib/aws-stepfunctions";
+import { Choice, Succeed, Map, StateMachine, Condition, TaskInput, CustomState, LogLevel } from "aws-cdk-lib/aws-stepfunctions";
 import { LambdaInvoke } from "aws-cdk-lib/aws-stepfunctions-tasks";
 import { WebCrawlerSteps } from './web-crawler-step-lambdas';
 import { WEB_CRAWLER_STATE_MACHINE_NAME } from './constants';
+import {Bucket} from "aws-cdk-lib/aws-s3";
+import { PolicyStatement, Effect } from 'aws-cdk-lib/aws-iam';
+import {LogGroup} from 'aws-cdk-lib/aws-logs';
```
Comment (cogwirrel) [inline src/constructs/webcrawler/web-crawler-state-machine.ts]: Nit: please match code style :)

File: src/constructs/webcrawler/web-crawler-state-machine.ts [hunk 2]
```diff
@@ -20,9 +25,10 @@ export default class WebCrawlerStateMachine extends Construct {
     super(scope, id);
 
     // Create tasks for each of the web crawler steps
-    const crawlPageAndQueueUrls = new LambdaInvoke(this, 'CrawlPageAndQueueUrls', {
-      lambdaFunction: props.steps.crawlPageAndQueueUrls,
-    });
+    // const crawlPageAndQueueUrls = new LambdaInvoke(this, 'CrawlPageAndQueueUrls', {
+    //  lambdaFunction: props.steps.crawlPageAndQueueUrls,
+    //});
```
Comment (cogwirrel) [inline src/constructs/webcrawler/web-crawler-state-machine.ts]: Let's get rid of all the commented code :)

File: src/lambda/config/constants.ts [hunk 3]
```diff
@@ -1,27 +1,41 @@
...
-export const PARALLEL_URLS_TO_SYNC = 200;
+const ENV_URLS_TO_SYNC = parseInt(process.env.PARALLEL_URLS_TO_SYNC || '')
+export const PARALLEL_URLS_TO_SYNC = Number.isInteger(ENV_URLS_TO_SYNC) ? ENV_URLS_TO_SYNC: 1000;
 
 /**
  * Provisioned write capacity for each context table.
  *
  * The write capacity should be adjusted with the PARALLEL_URLS_TO_SYNC to avoid throttling errors. Context tables are
  * ephemeral, existing only for the duration of a given web crawl execution.
+ * @deprecated use on-demand capacity to let DynamoDB handle the scaling instead of provisioned capacity
```
Comment (cogwirrel) [inline src/lambda/config/constants.ts]: Let's delete these completely as pay per request is much better for this use case :)

File: src/constructs/webcrawler/web-crawler-state-machine.ts [hunk 4]
```diff
@@ -34,6 +40,75 @@ export default class WebCrawlerStateMachine extends Construct {
...
+    // As of Oct 2023, CDK does not support Distributed Map state yet
+    // See: https://github.com/aws/aws-cdk/issues/23216
+    // Using CustomState as a workaround
+    //const dummyMap = new Map(this, "DummyMap");
+    //dummyMap.iterator(crawlPageAndQueueUrls).next(readQueuedUrls)
+    const createDistMapForEachQueuedUrl = () => {
+      return new CustomState(this, "ForEachQueuedUrl", {
+        stateJson: {
+          "Type": "Map",
+          "MaxConcurrency": 1000,
```
Comment (cogwirrel) [inline src/constructs/webcrawler/web-crawler-state-machine.ts]: Can we use the `PARALLEL_URLS_TO_SYNC` constant?

File: src/constructs/webcrawler/web-crawler-state-machine.ts [hunk 5]
```diff
@@ -45,20 +120,44 @@ export default class WebCrawlerStateMachine extends Construct {
...
     // Grant the startCrawling lambda permissions to start an execution of this state machine
     webCrawlerStateMachine.grantStartExecution(props.steps.startCrawl);
+    // Grant the webcrawler state machine permissions to read and write to the working bucket
+    props.workingBucket.grantReadWrite(webCrawlerStateMachine.role);
+    
+    // manually add permission to execute own state machine, required by Distributed Map state
+    // for child executions. 
+    webCrawlerStateMachine.addToRolePolicy(new PolicyStatement({
+      effect: Effect.ALLOW,
+      actions: ["states:*"],
```
Comment (cogwirrel) [inline src/constructs/webcrawler/web-crawler-state-machine.ts]: Can we scope this down to just the required actions?

File: src/constructs/webcrawler/web-crawler-step-lambdas.ts [hunk 6]
```diff
@@ -69,7 +72,10 @@ export default class WebCrawlerStepLambdas extends Construct {
     // Lambda for reading queued urls from the context table
     const readQueuedUrls = buildLambda('ReadQueuedUrlsLambda', 'readQueuedUrlsHandler');
     props.historyTable.grantReadWriteData(readQueuedUrls);
+    props.workingBucket.grantReadWrite(readQueuedUrls);
     readQueuedUrls.addToRolePolicy(props.createContextTablePolicy(['Query']));
+    readQueuedUrls.addEnvironment("STATE_MACHINE_URL_THRESHOLD", "100000");
+    readQueuedUrls.addEnvironment("PARALLEL_URLS_TO_SYNC", "1000");
```
Comment (cogwirrel) [inline src/constructs/webcrawler/web-crawler-step-lambdas.ts]: I like having the constants loaded by the lambdas via environment variables, but can we still have the definitions be in a constants file which explains what the constants to make it a bit easier for users to find how to adjust these?

I think it'd be best to:

* Introduce a new file `src/constants.ts` which has the comments about the constants and what they mean, as well as the values used in the samples.
* Update the lambdas to read directly from the env - fine if you want to abstract this away as in the `src/lambdas/constants.ts` file but let's call it `src/lambdas/env.ts` instead.
* Update the CDK code to use the constants defined in `src/constants.ts`

File: src/lambda/steps/2_readQueuedUrls.ts [hunk 7]
```diff
@@ -30,13 +38,27 @@ export const readQueuedUrls = async (crawlContext: CrawlContext) => {
     batchUrlCount: totalBatchUrlCount,
   });
 
+  let queuedPaths = urlsToVisit.map((path) => ({
```
Comment (cogwirrel) [inline src/lambda/steps/2_readQueuedUrls.ts]: Nit: Prefer `const` where you don't reassign variables :)

File: src/lambda/steps/2_readQueuedUrls.ts [hunk 8]
```diff
@@ -30,13 +38,27 @@ export const readQueuedUrls = async (crawlContext: CrawlContext) => {
...
+    path,
+    crawlContext
+  }))
+
+  // Save queuedPaths to S3
+  const folder = 'temp-queued-paths';
+  const key = `${folder}/${crawlContext.crawlId}.queuedPaths.json`;
+  let result = await s3.putObject({ Bucket: WORKING_BUCKET, Key: key, Body: JSON.stringify(queuedPaths) }).promise();
```
Comment (cogwirrel) [inline src/lambda/steps/2_readQueuedUrls.ts]: Nit: `const`

File: src/lambda/steps/3_crawlPageAndQueueUrls.ts [hunk 9]
```diff
@@ -51,7 +51,7 @@ export const crawlPageAndQueueUrls = async (
     }, destination);
     console.log('Synced content from', path);
 
-    console.log('Queueing', urlPaths.length, 'new urls to visit', urlPaths);
+    console.log('Queueing ', urlPaths.length, ' new urls to visit', urlPaths);
```
Comment (cogwirrel) [inline src/lambda/steps/3_crawlPageAndQueueUrls.ts]: The spaces are rendered automatically by `console.log` so you shouldn't need to add them :)

File: src/lambda/utils/contextTable.ts [hunk 10]
```diff
@@ -46,10 +46,12 @@ export const createContextTable = async (target: CrawlInputWithId, contextTableN
         KeyType: 'RANGE',
       },
     ],
+    BillingMode: 'PAY_PER_REQUEST'
+    /*,
```
Comment (cogwirrel) [inline src/lambda/utils/contextTable.ts]: Again let's get rid of commented code :)

File: src/lambda/utils/pagination.ts [hunk 11]
```diff
@@ -35,6 +35,9 @@ export const paginatedRequest = async <Params>(
     }
     token = response[nextTokenKey];
   } while (token);
+  //
+  console.log('paginatedRequest.results.length: ', results.length)
```
Comment (cogwirrel) [inline src/lambda/utils/pagination.ts]: Let's get rid of this :)

File: src/stacks/web-crawler-stack.ts [hunk 12]
```diff
@@ -67,10 +70,11 @@ export class WebCrawlerStack extends Stack {
       createContextTablePolicy,
       kendra: props.kendra,
       historyTable,
+      workingBucket,
       webCrawlerStateMachineArn,
     });
 
     // Create the state machine
-    new WebCrawlerStateMachine(this, 'WebCrawlerStateMachine', { steps });
+    const webCrawlerStateMachine = new WebCrawlerStateMachine(this, 'WebCrawlerStateMachine', { steps, workingBucket, webCrawlerStateMachineArn});
```
Comment (cogwirrel) [inline src/stacks/web-crawler-stack.ts]: Nit: I think we can delete the variable as it's unused :)

File: src/constructs/webcrawler/web-crawler-state-machine.ts [hunk 13]
```diff
@@ -45,20 +120,44 @@ export default class WebCrawlerStateMachine extends Construct {
             // Continue crawling in another state machine execution
             continueExecution.next(new Succeed(this, 'ContinuingInAnotherExecution')))
             // Crawl every page we read from the queue in parallel
-            .otherwise(new Map(this, 'ForEachQueuedUrl', { itemsPath: '$.Payload.queuedPaths' })
-              .iterator(crawlPageAndQueueUrls).addCatch(readQueuedUrls)
-              .next(readQueuedUrls))
+            //.otherwise(new Map(this, 'ForEachQueuedUrl', { itemsPath: '$.Payload.queuedPaths' })
+            //  .iterator(crawlPageAndQueueUrls).addCatch(readQueuedUrls)
+            //  .next(readQueuedUrls))
```
Comment (cogwirrel) [inline src/constructs/webcrawler/web-crawler-state-machine.ts]: Let's get rid of this :)

File: src/constructs/webcrawler/web-crawler-state-machine.ts [hunk 14]
```diff
@@ -45,20 +120,44 @@ export default class WebCrawlerStateMachine extends Construct {
...
+            // Change from Inline Map to Distributed Map
```
Comment (cogwirrel) [inline src/constructs/webcrawler/web-crawler-state-machine.ts]: Let's delete this too, users can look up the git history if they want to know what has changed :)

File: src/constructs/webcrawler/web-crawler-state-machine.ts [hunk 15]
```diff
@@ -34,6 +40,75 @@ export default class WebCrawlerStateMachine extends Construct {
       lambdaFunction: props.steps.continueExecution,
     });
 
+    // As of Oct 2023, CDK does not support Distributed Map state yet
+    // See: https://github.com/aws/aws-cdk/issues/23216
+    // Using CustomState as a workaround
+    //const dummyMap = new Map(this, "DummyMap");
+    //dummyMap.iterator(crawlPageAndQueueUrls).next(readQueuedUrls)
```
Comment (cogwirrel) [inline src/constructs/webcrawler/web-crawler-state-machine.ts]: Nit: let's remove the commented code :)

File: bin/build-chrome-layer.sh [hunk 16]
```diff
@@ -6,4 +6,5 @@
 set -e
 
 mkdir -p chromelayer/nodejs/node_modules/
-cd chromelayer/nodejs && npm install chrome-aws-lambda@10.1.0 puppeteer-core@10.1.0 --no-bin-links --no-optional --no-package-lock --no-save --no-shrinkwrap && cd -
+#cd chromelayer/nodejs && npm install chrome-aws-lambda@10.1.0 puppeteer-core@10.1.0 --no-bin-links --no-optional --no-package-lock --no-save --no-shrinkwrap && cd -
+cd chromelayer/nodejs && npm install chrome-aws-lambda@10.1.0 puppeteer-core@10.1.0 --no-bin-links --no-package-lock --no-save --no-shrinkwrap && cd -
```
Comment (cogwirrel) [inline bin/build-chrome-layer.sh]: Let's get rid of the commented command - I'm interested to know why you changed this? :)

**Review by cogwirrel — CHANGES_REQUESTED** [review]
Thanks so much for this! Distributed map will be a massive improvement - awesome! 😄 

PR looks great! Just a few minor comments :)

File: src/constructs/webcrawler/web-crawler-state-machine.ts (hunk 4 above)
Comment (cogwirrel) [reply src/constructs/webcrawler/web-crawler-state-machine.ts]: Discussed offline - please add comment explaining this vs parallel urls to sync :)

File: src/constructs/webcrawler/web-crawler-state-machine.ts (hunk 1 above)
Comment (philiptran) [reply src/constructs/webcrawler/web-crawler-state-machine.ts]: Fixed

File: src/constructs/webcrawler/web-crawler-state-machine.ts (hunk 2 above)
Comment (philiptran) [reply src/constructs/webcrawler/web-crawler-state-machine.ts]: Fixed

File: src/lambda/config/constants.ts (hunk 3 above)
Comment (philiptran) [reply src/lambda/config/constants.ts]: Removed the src/lambda/config/constants.ts and added the constants to src/constructs/webcrawler/constants.ts. These serve as default values for the environment variables of the Lambda steps.

File: src/constructs/webcrawler/web-crawler-state-machine.ts (hunk 4 above)
Comment (philiptran) [reply src/constructs/webcrawler/web-crawler-state-machine.ts]: Added comments explaining these constants in src/constructs/webcrawler/constants.ts

File: src/constructs/webcrawler/web-crawler-state-machine.ts (hunk 5 above)
Comment (philiptran) [reply src/constructs/webcrawler/web-crawler-state-machine.ts]: Fixed

File: src/constructs/webcrawler/web-crawler-step-lambdas.ts (hunk 6 above)
Comment (philiptran) [reply src/constructs/webcrawler/web-crawler-step-lambdas.ts]: Implemented the following:
- Deleted `src/lambda/config/constants.ts` as the constants will be loaded by the Lambdas via environment variables
- Added the constants to `src/constructs/webcrawler/constants.ts` with comments explaining these constants.
- Updated CDK code to use these constants as default values for the Lambdas' environment variables.
- Added `src/lambda/utils/env.ts` and `getEnvVariableAsInteger` util function.

File: src/lambda/steps/2_readQueuedUrls.ts (hunk 7 above)
Comment (philiptran) [reply src/lambda/steps/2_readQueuedUrls.ts]: Fixed

File: src/lambda/steps/2_readQueuedUrls.ts (hunk 8 above)
Comment (philiptran) [reply src/lambda/steps/2_readQueuedUrls.ts]: Fixed

File: src/lambda/steps/3_crawlPageAndQueueUrls.ts (hunk 9 above)
Comment (philiptran) [reply src/lambda/steps/3_crawlPageAndQueueUrls.ts]: Fixed

File: src/lambda/utils/contextTable.ts (hunk 10 above)
Comment (philiptran) [reply src/lambda/utils/contextTable.ts]: Fixed

File: src/lambda/utils/pagination.ts (hunk 11 above)
Comment (philiptran) [reply src/lambda/utils/pagination.ts]: Fixed.

File: src/stacks/web-crawler-stack.ts (hunk 12 above)
Comment (philiptran) [reply src/stacks/web-crawler-stack.ts]: Fixed.

File: src/constructs/webcrawler/web-crawler-state-machine.ts (hunk 13 above)
Comment (philiptran) [reply src/constructs/webcrawler/web-crawler-state-machine.ts]: Fixed.

File: src/constructs/webcrawler/web-crawler-state-machine.ts (hunk 14 above)
Comment (philiptran) [reply src/constructs/webcrawler/web-crawler-state-machine.ts]: Fixed

File: src/constructs/webcrawler/web-crawler-state-machine.ts (hunk 15 above)
Comment (philiptran) [reply src/constructs/webcrawler/web-crawler-state-machine.ts]: Fixed

File: bin/build-chrome-layer.sh (hunk 16 above)
Comment (philiptran) [reply bin/build-chrome-layer.sh]: Reverted the change as the `--no-optional` option here was not the cause of my local env issue. See image below.

Image (1)

**Comment by philiptran** [comment]
@cogwirrel thank you for the detailed comments. Pushed in 2 new commits addressing these comments. Appreciate your review.

**Review by cogwirrel — APPROVED** [review]
Thanks very much Philip! 🚀🚀🚀

**Pull request merged**